feat: `rdme docs upload`

kanadgupta opened the pull request:
| 🚥 Resolves RM-11796, RM-11797 |
| :------------------- |

## 🧰 Changes

### 🚧 outstanding work

- [x] core command + helper logic
- [x] tests
- [x] loading states
- [x] complete error messages
- [x] docs 
   - [x] migration guide (see https://github.com/readmeio/rdme/pull/1161)
   - [x] callouts in `documentation/rdme.md` (see https://github.com/readmeio/rdme/pull/1163)


### 🔜 to be addressed in follow-up PRs

- [ ] safeguards for users that use bidirectional syncing (ticketed in RM-11883 and RM-11901)
- [ ] updating the v9 migration guide + deprecation warnings (ticketed in RM-11902)

### things to call out

- on a high level, we're mostly matching the functionality of v9's `rdme docs` in that we're taking markdown files and sending their contents as JSON payloads to the ReadMe API. inferring page hierarchy from directory/file structure is **out of scope** for this work.
- one big change from `rdme@9` is that we now validate page frontmatter against the schemas from our OAS before making any API requests and prompt the user to migrate any frontmatter that we recognize as legacy APIv1 attributes. i'm open to reevaluating this paradigm, but i think it'll be helpful for folks migrating. as an interim escape hatch, this command has a hidden `--skip-validation` flag, just in case.
- our API expects category and parent page URIs to be a full URI (e.g., `/versions/stable/categories/guides/some-uri`), but that's a bit arduous for `rdme` users since most of the URI can be inferred from the `rdme` command context. as a result, `rdme` will automatically expand any category/parent URIs that don't match that pattern (e.g., a category URI frontmatter value of `some-uri` will be changed to `/versions/stable/categories/guides/some-uri` before making the `POST`/`PATCH` request. the frontmatter will still say `some-uri`.)
- the vast majority of this diff is tests + fixtures + our OAS. the three files of note are:
   - `src/lib/syncPagePath.ts`
   - `src/lib/frontmatter.ts`
   - `src/commands/docs/upload.ts`
- i've tried to build this in a testable and robust way where we can easily add support for reference pages, custom pages, etc. in the future
- `PATCH` requests to our API (i.e., for making updates to guide pages) that include `slug` in the body payload will always rename the page. so if you have a slug called `some-page` and your body payload contains `some-page`, the API will change your slug to `some-page-1`. this is most likely unexpected behavior for users of `rdme`, so i'm **always** removing the slug from the body payload in favor of using the URL instead. the trade-off is that page renaming isn't really a thing via `rdme`, but that's why we support bidirectional syncing, right?

## 🧬 QA & Testing

we're at nearly 100% test coverage with this work 📈 to QA it yourself, check out this branch and run the following:

```sh
# install + set up 
npm ci
npm run build
```

```sh
# grab an API key from a readme-refactored-enabled project and pop it at the end of this command:
bin/run.js docs upload __tests__/__fixtures__/docs/mixed-docs --key <key>
```

kanadgupta committed:
chore: deprecate syncDocsPath

kanadgupta committed:
Merge branch 'next' into kanad-2025-01-14/docs-v2

kanadgupta committed:
refactor: cleaner arg definitions

kanadgupta committed:
chore: smarter body processing

kanadgupta committed:
Merge branch 'next' into kanad-2025-01-14/docs-v2

kanadgupta committed:
Merge branch 'next' into kanad-2025-01-14/docs-v2

kanadgupta committed:
refactor: rename readDoc to readPage

kanadgupta committed:
chore: deprecation warnings for APIv1 helpers

kanadgupta committed:
chore: smol type refactor

kanadgupta committed:
chore: deprecate field

kanadgupta committed:
chore: smol type improvement

kanadgupta committed:
chore: remove unused fixture

kanadgupta committed:
fix: bypass graymatter cache

kanadgupta committed:
refactor: move `readPage` helper function's `this`

kanadgupta committed:
chore: remove unused snapshots

kanadgupta committed:
chore: remove a few more unused fixtures

kanadgupta committed:
chore(migration-guide): small whitespace change

kanadgupta committed:
chore: remove more unused fixtures

kanadgupta committed:
chore: remove even more fixtures

kanadgupta committed:
docs: changelog copy tweaks

kanadgupta committed:
feat(docs): first pass at APIv2 command

kanadgupta committed:
chore: re-indent snapshots

kanadgupta committed:
Merge branch 'next' into kanad-2025-01-14/docs-v2

kanadgupta committed:
chore: fix merge conflicts

kanadgupta committed:
chore: reduce whitespace where possible

kanadgupta committed:
revert: bring back whitespace

kanadgupta committed:
chore: jsdoc that no longer applies

kanadgupta committed:
chore: move fetch calls further down tree

kanadgupta committed:
refactor: move fetchSchema/fetchMappings into fetch file

kanadgupta committed:
chore: jsdoc

kanadgupta committed:
chore: move oasfetchmock to different file

kanadgupta committed:
test: more assertions

kanadgupta committed:
fix: copy edits, loading states

kanadgupta committed:
chore: lint

kanadgupta committed:
test: add assertion that source URL header is being sent properly

kanadgupta committed:
chore: rearrange tests

kanadgupta committed:
refactor: type enhancement

kanadgupta committed:
test: add dry run case

kanadgupta committed:
test: reorg

kanadgupta committed:
test: dry run single failure test

kanadgupta committed:
chore: reformat api errors

kanadgupta committed:
feat: first pass at proper error messages

kanadgupta committed:
refactor: dry up the logging logic

kanadgupta committed:
chore: remove mostly useless log

kanadgupta committed:
chore: emoji tweaks

kanadgupta committed:
chore: fix tests in GHA runner

kanadgupta committed:
fix: consistent error response copy

kanadgupta committed:
test: add test for category mapping

erunion commented on src/lib/syncPagePath.ts:
```diff
@@ -0,0 +1,411 @@
+ * Reads the contents of the specified Markdown or HTML file
+ * and creates/updates the corresponding page in ReadMe
+ */
+async function pushPage(
```
Might be worth renaming this and `PageMetadata` to be about a "file" instead of a "page" so it's a little clearer as the only thing we really use the "page" nomenclature around is are custom pages.

kanadgupta replied:
ah so i actually intentionally went with "page" since that's the nomenclature we use in APIv2 code for guides and reference pages. "file" felt a bit too vague since this is only concerned with markdown and not openapi

kanadgupta committed:
refactor: use AggregateError

h/t @erunion

erunion commented on src/lib/syncPagePath.ts:
```diff
@@ -0,0 +1,413 @@
+    return edges;
+  }, []);
+
+  return toposort.array(files, dependencies);
```
i follow what `toposort` is doing from their docs, but how come it's being used here? to ensure that we create parents before children?

mind adding some comments?

kanadgupta replied:
can do! for the record this is carried over from @kellyjosephprice's work over in https://github.com/readmeio/rdme/pull/973 and i've blindly trusted it ever since lol

kanadgupta replied:
https://github.com/readmeio/rdme/pull/1159/commits/8153b90bf512eb65b4a8338fc0445a4e28796f72

erunion commented on src/lib/syncPagePath.ts:
```diff
@@ -0,0 +1,413 @@
+        this.warn(
+          `${filesWithUnfixableIssues.length} file(s) have issues that cannot be fixed automatically. The upload will proceed but we recommend addressing these issues. Please get in touch with us at support@readme.io if you need a hand.`,
+        );
+
+        //
```
```suggestion
```

kanadgupta replied:
whoops

kanadgupta replied:
https://github.com/readmeio/rdme/pull/1159/commits/8153b90bf512eb65b4a8338fc0445a4e28796f72

erunion reviewed (commented):
leaving formal approval off until this is ready for review but aside from a couple minor questions this looks great. awesome job.

kanadgupta commented on __tests__/lib/frontmatter.test.ts:
```diff
@@ -0,0 +1,340 @@
+    expect(result.updatedData).toMatchInlineSnapshot(`
+      {
+        "category": {
+          "uri": "some-category-slug",
```
yeah, it's just raw slugs from the APIv1 data 🙃

erunion replied:
feel like i remember aaron and i talking about this and i had the same question. fine with this as-is i guess but it's definitely weird

kanadgupta commented on src/lib/frontmatter.ts:
```diff
@@ -0,0 +1,155 @@
+const addFormats = _addFormats as unknown as typeof _addFormats.default;
+
+/**
+ * Validates the front matter data, fixes any issues, and returns the results.
```
oh yeah i've always went with jekyll since they're the OG and their approach is "front matter" but don't feel super strongly about it. will change!

kanadgupta replied:
https://github.com/readmeio/rdme/pull/1159/commits/8153b90bf512eb65b4a8338fc0445a4e28796f72

kanadgupta commented on src/lib/readmeAPIFetch.ts:
```diff
@@ -436,3 +446,57 @@ export function cleanAPIv1Headers(
+  }
+
+  this.debug(`unable to find parse out schema for ${JSON.stringify(oas)}`);
+  return readmeAPIv2Oas.paths[`/versions/{version}/${this.route}/{slug}`].patch.requestBody.content['application/json']
```
yeah, agreed with that. truthfully i'd much prefer the one from the API but i wasn't sure if i'd get around to actually figuring out https://github.com/readmeio/readme/pull/13807 lol 😭

kanadgupta replied:
thoughts on me leaving this as is for now and removing it once we get https://github.com/readmeio/readme/pull/13807 out the door?

erunion replied:
sounds good

kanadgupta committed:
chore: feedback

kanadgupta committed:
chore: frontmatter, not front matter

kanadgupta committed:
Merge branch 'next' into kanad-2025-01-14/docs-v2

kanadgupta committed:
docs: whoops

kanadgupta commented on src/commands/docs/upload.ts:
```diff
@@ -0,0 +1,71 @@
+    github: githubFlag,
+    key: keyFlag,
+    'dry-run': Flags.boolean({
+      description: 'Runs the command without creating/updating any changelogs in ReadMe. Useful for debugging.',
```
https://github.com/readmeio/rdme/pull/1159/commits/8153b90bf512eb65b4a8338fc0445a4e28796f72 and c9e19fba3f30409794145395345f79696400baaf

kanadgupta commented on src/commands/docs/upload.ts:
```diff
@@ -0,0 +1,71 @@
+  static aliases = [
+    // 'docs',
+    // this needs to be separated by a colon in order for autocomplete to work properly
+    'guides:upload',
```
https://github.com/readmeio/rdme/pull/1159/commits/8153b90bf512eb65b4a8338fc0445a4e28796f72

kanadgupta commented on package.json:
```diff
@@ -47,6 +47,9 @@
     "@oclif/plugin-help": "^6.2.15",
     "@oclif/plugin-not-found": "^3.2.28",
     "@oclif/plugin-warn-if-update-available": "^3.1.19",
+    "@readme/better-ajv-errors": "^2.0.0",
```
https://github.com/readmeio/rdme/pull/1162

kanadgupta commented on __tests__/commands/docs/upload.test.ts:
```diff
@@ -0,0 +1,451 @@
+  });
+
+  beforeEach(() => {
+    fs.writeFileSync = vi.fn(() => {});
```
https://github.com/readmeio/rdme/pull/1159/commits/8153b90bf512eb65b4a8338fc0445a4e28796f72

kanadgupta committed:
chore: rely on top-level setup file for nock setup

kanadgupta commented on __tests__/commands/docs/upload.test.ts:
```diff
@@ -0,0 +1,451 @@
+  let run: (args?: string[]) => OclifOutput;
+
+  beforeAll(() => {
+    nock.disableNetConnect();
```
c5261812661d6fcef64b6dc91b3b0e7f3dd09b4c

kanadgupta committed:
fix: delete parentDoc if no mapping is available

kanadgupta committed:
Merge branch 'next' into kanad-2025-01-14/docs-v2

kanadgupta committed:
Merge branch 'next' into kanad-2025-01-14/docs-v2

kanadgupta committed:
docs: migration guide updates for `rdme docs upload` (#1161)

| 🚥 Resolves RM-11800 |
| :------------------- |

## 🧰 Changes

first (and hopefully totally adequate!) pass at the `rdme docs`
migration guide. we can continue to iterate after this initial beta is
out, based on feedback!

## 🧬 QA & Testing

does this language all look sound to you? any big points that we're
missing?

kanadgupta committed:
docs: autogenerate command docs

erunion commented on __tests__/lib/frontmatter.test.ts:
```diff
@@ -0,0 +1,337 @@
+  });
+});
+
+describe('#fetchSchema', () => {
```
May want to move this test case to a different test file because `fetchSchema` is part of `readmeAPIFetch`, not `frontmatter`.

kanadgupta replied:
ah yeah good catch, ty

erunion commented on src/lib/frontmatter.ts:
```diff
@@ -0,0 +1,157 @@
+) {
+  this.debug(`writing fixes to ${metadata.filePath}`);
+  const result = grayMatter.stringify(metadata.content, updatedData);
+  fs.writeFileSync(metadata.filePath, result, { encoding: 'utf-8' });
```
Wonder if we need to do some detection on these files to preserve their existing encoding. I know that `latin-1` doesn't support emojis, and saving an emoji into one results in some garbage text, wonder it converting that then to UTF-8 would fix that emoji or add more garbage.

erunion replied:
Ran a small test and we thankfully don't do any corruption on it so this should be fine.

```ts
const fs = require('node:fs');

const filePath = 'debug.md';
const content = `---
name: test
---
shrimp is bugs 🦐`;

fs.writeFileSync(filePath, content, { encoding: 'latin1' });

const file = fs.readFileSync(filePath, { encoding: 'utf-8' });
fs.writeFileSync(filePath, file, { encoding: 'utf-8' });
```

```
---
name: test
---
shrimp is bugs >�
```

erunion commented on src/lib/syncPagePath.ts:
```diff
@@ -0,0 +1,418 @@
+      return this.readmeAPIFetch(
+        route,
+        { method: 'POST', headers, body: JSON.stringify(payload) },
+        { filePath, fileType: 'path' },
```
What's this third argument for?

kanadgupta replied:
this:

https://github.com/readmeio/rdme/blob/9f84783d19556032ec9e2b911be20630a657b408/src/lib/readmeAPIFetch.ts#L264-L278

erunion reviewed (approved):
**mildly** concerned about some file encoding bullshit but it's nothing that can't wait. 

this all looks awesome.

kanadgupta committed:
chore: move tests to correct spot

kanadgupta commented:
:tada: This PR is included in version 10.2.0-next.1 :tada:

The release is available on:
- [npm package (@next dist-tag)](https://www.npmjs.com/package/rdme/v/10.2.0-next.1)
- `v10.2.0-next.1`
- [GitHub release](https://github.com/readmeio/rdme/releases/tag/v10.2.0-next.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

kanadgupta commented:
:tada: This PR is included in version 10.2.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/rdme/v/10.2.0)
- `v10.2.0`
- [GitHub release](https://github.com/readmeio/rdme/releases/tag/v10.2.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
